clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -274,10 +273,10 @@ public ReadyCheckResult ready(Cluster cluster, long nowMs) {
             Deque<RecordBatch> deque = entry.getValue();
 
             Node leader = cluster.leaderFor(part);
-            if (leader == null) {
-                unknownLeadersExist = true;
-            } else if (!readyNodes.contains(leader)) {
-                synchronized (deque) {
+            synchronized (deque) {
+                if (leader == null && !deque.isEmpty()) {
```

Comment:
The fact that this check is necessary makes me realize that `this.batches` will also grow indefinitely, right? It looks like we don't ever remove entries from that map. Not necessarily something we need to address in this patch, but might be another thing we should address, especially for something like the REST proxy where this could grow quite large, requiring every `ready` check to traverse the entire list, acquiring 1 lock per `TopicPartition` since this patch requires moving the lock acquisition before checking if we've already added the leader for the `TopicPartition` to the set of ready nodes.